bin/pyrdp-convert.py

```diff
@@ -0,0 +1,433 @@
+OUTFILE_FORMAT = '{prefix}{timestamp}_{src}-{dst}.{ext}'
+
+class Mp4Layer(LayerChainItem):
+    def __init__(self, sink: Mp4EventHandler):
+        self.sink = sink
+        self.player = PlayerLayer()
+        self.player.addObserver(sink)
+
+    def sendBytes(self, data: bytes):
+        self.player.recv(data)
+
+
+class CustomMITMRecorder(MITMRecorder):
+    currentTimeStamp: int = None
+
+    def getCurrentTimeStamp(self) -> int:
+        return self.currentTimeStamp
+
+    def setTimeStamp(self, timeStamp: int):
+        self.currentTimeStamp = timeStamp
+
+
+class RDPReplayerConfig(MITMConfig):
+    @property
+    def replayDir(self) -> Path:
+        return self.outDir
+
+    @property
+    def fileDir(self) -> Path:
+        return self.outDir
+
+
+class RDPReplayer(RDPMITM):
+    def __init__(self, output_path: str, mp4: bool = False):
+        def sendBytesStub(_: bytes):
+            pass
+
+        output_path = Path(output_path)
+        output_directory = output_path.absolute().parent
+
+        logger = logging.getLogger(LOGGER_NAMES.MITM_CONNECTIONS)
+        log = SessionLogger(logger, "replay")
+
+        config = RDPReplayerConfig()
+        config.outDir = output_directory
+        # We'll set up the recorder ourselves
+        config.recordReplays = False
+
+        transport = FileLayer(output_path)
+        state = RDPMITMState(config)
+
+        sink = None
+        if mp4:
+            sink = Mp4EventHandler(output_path)
+            transport = Mp4Layer(sink)
+
+        super().__init__(log, log, config, state, CustomMITMRecorder([transport], state))
+
+        self.client.tcp.sendBytes = sendBytesStub
+        self.server.tcp.sendBytes = sendBytesStub
+        self.state.useTLS = True
+
+    def start(self):
+        pass
+
+    def recv(self, data: bytes, from_client: bool):
+        if from_client:
+            self.client.tcp.dataReceived(data)
+        else:
+            self.server.tcp.dataReceived(data)
+
+    def setTimeStamp(self, timeStamp: float):
+        self.recorder.setTimeStamp(int(timeStamp * 1000))
+
+    def connectToServer(self):
+        pass
+
+    def startTLS(self):
+        pass
+
+    def sendPayload(self):
+        pass
+
+
+def tcp_both(p) -> str:
+    """Session extractor which merges both sides of a TCP channel."""
+
+    if 'TCP' in p:
+        return str(sorted(['TCP', p[IP].src, p[TCP].sport, p[IP].dst, p[TCP].dport], key=str))
+    return 'Other'
+
+
+def findClientRandom(stream: PacketList, limit: int = 10) -> str:
+    """Find the client random offset and value of a stream."""
+    for n, p in enumerate(stream):
+        if n >= limit:
+            return ''  # Didn't find client hello.
+        try:
+            tls = TLS(p.load)
+            hello = tls.msg[0]
+            if not isinstance(hello, TLSClientHello):
+                continue
+            return hexlify(pkcs_i2osp(hello.gmt_unix_time, 4) + hello.random_bytes).decode()
+        except Exception:
```

Comment:
Because I couldn't find documentation on what exceptions `TLS()` throws and I don't care what it throws: If that packet cannot make it to the `isinstance()` check, it's definitely not ClientHello, and if anything on line `135` fails, we can't get the secret, so the outcome doesn't change: We can't decrypt the trace.